miniapps/autodiff/taddensemat.hpp

```diff
@@ -0,0 +1,513 @@
+  from AD to hand-optimized code as it does not require a change
+  in the interface or the code structure. TADDenseMatrix is intended
+  to be utilized for dense serial matrices. The objects can be combined
+  with TADVector or standard Vector.*/
+template<typename dtype>
+class TADDenseMatrix
+{
+private:
+   int height; ///< Dimension of the output / number of rows in the matrix.
+   int width;  ///< Dimension of the input / number of columns in the matrix.
+   dtype *data;
+   int capacity; // zero or negative capacity means we do not own the data.
+
+public:
+   /// Get the height (size of output) of the Operator. Synonym with NumRows().
+   inline int Height() const { return height; }
+   /** @brief Get the number of rows (size of output) of the Operator. Synonym
+       with Height(). */
+   inline int NumRows() const { return height; }
+
+   /// Get the width (size of input) of the Operator. Synonym with NumCols().
+   inline int Width() const { return width; }
+   /** @brief Get the number of columns (size of input) of the Operator. Synonym
+       with Width(). */
+   inline int NumCols() const { return width; }
+
+   /** Default constructor for TADDenseMatrix.
+      Sets data = NULL and height = width = 0. */
+   TADDenseMatrix()
+   {
+      data = nullptr;
+      capacity = 0;
+      height = 0;
+      width = 0;
+   }
+
+   /// Copy constructor
+   template<typename idtype>
+   TADDenseMatrix(const TADDenseMatrix<idtype> &m)
+   {
+      height = m.GetHeight();
+      width = m.GetWidth();
+      const int hw = height * width;
+      if (hw > 0)
+      {
+         idtype *mdata = m.Data();
+         MFEM_ASSERT(mdata, "invalid source matrix");
+         data = new dtype[hw];
+         capacity = hw;
+         for (int i = 0; i < hw; i++)
+         {
+            data[i] = mdata[i];
+         }
+      }
+      else
+      {
+         data = nullptr;
+         capacity = 0;
+         width = 0;
+         height = 0;
+      }
+   }
+   /// Copy constructor using standard DenseMatrix
+   TADDenseMatrix(const DenseMatrix &m)
+   {
+      height = m.Height();
+      width = m.Width();
+      const int hw = height * width;
+      if (hw > 0)
+      {
+         double *mdata = m.Data();
+         MFEM_ASSERT(mdata, "invalid source matrix");
+         data = new dtype[hw];
+         capacity = hw;
+         for (int i = 0; i < hw; i++)
+         {
+            data[i] = mdata[i];
+         }
+      }
+      else
+      {
+         data = nullptr;
+         capacity = 0;
+         width = 0;
+         height = 0;
+      }
+   }
+
+   /// Creates square matrix of size s.
+   explicit TADDenseMatrix(int s)
+   {
+      MFEM_ASSERT(s >= 0, "invalid DenseMatrix size: " << s);
+      height = s;
+      width = s;
+      capacity = s * s;
+      if (capacity > 0)
+      {
+         data = new dtype[capacity](); // init with zeroes
+      }
+      else
+      {
+         data = NULL;
```

Comment:
`nullptr`

Reply:
Is the goal here to check for overflow or just a zero size? In the case of `s * s` overflowing, I think throwing an exception would be better than just setting the data pointer to `nullptr`. Also, it seems if `data` is being set to nullptr, shouldn't height and width also be set to zero.

Reply:
Modified.

I have tried to replicate exactly the behavior of the  original DenseMatrix class in order to avoid any incompatibilities.